File: engine/runtime/src/main/scala/org/enso/compiler/ImportExportCache.scala
package org.enso.compiler

import org.enso.pkg.{QualifiedName, SourceFile}
import buildinfo.Info
import com.oracle.truffle.api.TruffleFile
import io.circe.generic.JsonCodec
import io.circe.parser._
import io.circe.syntax._
import org.bouncycastle.util.encoders.Hex
import org.enso.compiler.Cache.Roots
import org.enso.compiler.data.BindingsMap
import org.enso.editions.LibraryName
import org.enso.interpreter.runtime.EnsoContext

import java.nio.charset.{Charset, StandardCharsets}
import java.util.logging.Level
import scala.util.{Failure, Success, Try}

class ImportExportCache(library: LibraryName)
    extends Cache[ImportExportCache.CacheBindings] {
  import ImportExportCache._

  override type M = Metadata

  override val logLevel: Level    = Level.FINEST
  override def stringRepr: String = library.toString

  override def computeDigestFromSource(
    context: EnsoContext
  ): Option[String] = {
    context.getPackageRepository.getPackageForLibrary(library).map { pkg =>
      computeDigestOfLibrarySources(pkg.listSources())
    }
  }

  override def computeDigest(
    entry: ImportExportCache.CacheBindings
  ): Option[String] = {
    entry.sources.map(computeDigestOfLibrarySources)
  }

  private def computeDigestOfLibrarySources(
    pkgSources: List[SourceFile[TruffleFile]]
  ): String = {
    val digest = messageDigest()
    pkgSources.sortBy(_.qualifiedName)(NameOrdering).foreach { source =>
      digest.update(source.file.readAllBytes())
    }
    Hex.toHexString(digest.digest())
  }

  override def getCacheRoots(context: EnsoContext): Option[Cache.Roots] = {
    context.getPackageRepository.getPackageForLibrary(library).map { pkg =>
      val bindingsCacheRoot =
        pkg.getBindingsCacheRootForPackage(Info.ensoVersion)
      val localCacheRoot = bindingsCacheRoot.resolve(library.namespace)
      val distribution   = context.getDistributionManager
      val pathSegments = List(
        pkg.namespace,
        pkg.name,
        pkg.config.version,
        Info.ensoVersion,
        library.namespace
      )
      val path = distribution.LocallyInstalledDirectories.irCacheDirectory
        .resolve(pathSegments.mkString("/"))
      val globalCacheRoot = context.getTruffleFile(path.toFile)

      Roots(localCacheRoot, globalCacheRoot)
    }
  }

  override protected def extractObjectToSerialize(
    entry: CacheBindings
  ): AnyRef = {
    entry.bindings
  }

  override protected def validateReadObject(
    obj: AnyRef,
    meta: M
  ): Try[CacheBindings] =
    obj match {
      case bindings: MapToBindings =>
        Success(ImportExportCache.CacheBindings(library, bindings, None))
      case other =>
        Failure(
          new RuntimeException(
            s"unexpected deserialized object of ${other.getClass}"
          )
        )
    }

  override protected def metadataFromBytes(
    bytes: Array[Byte]
  ): Option[Metadata] = {
    val maybeJsonString = new String(bytes, metadataCharset)
    decode[Metadata](maybeJsonString).toOption
  }

  override protected def metadata(
    sourceDigest: String,
    blobDigest: String,
    entry: CacheBindings
  ): Array[Byte] = {
    val metadata = Metadata(sourceDigest, blobDigest)
    metadata.asJson.toString().getBytes(metadataCharset)
  }

  override protected def entryName: String = {
    library.name
  }

  override protected val dataSuffix: String = {
    bindingsCacheDataExtension
  }

  override protected def metadataSuffix: String = {
    bindingsCacheMetadataExtension
  }
}

object ImportExportCache {
  private val bindingsCacheDataExtension: String     = ".bindings"
  private val bindingsCacheMetadataExtension: String = ".bindings.meta"
  private val metadataCharset: Charset               = StandardCharsets.UTF_8

  @JsonCodec
  case class Metadata(
    sourceHash: String,
    blobHash: String
  ) extends Cache.Metadata

  case class CacheBindings(
    libraryName: LibraryName,
    bindings: MapToBindings,
    sources: Option[List[SourceFile[TruffleFile]]]
  )

  // Wrapper around map of bindings to avoid warnings about erasure
  // when matching the deserialized object
  case class MapToBindings(entries: Map[QualifiedName, BindingsMap])

  private object NameOrdering extends Ordering[QualifiedName] {
    override def compare(x: QualifiedName, y: QualifiedName): Int = {
      x.toString.compare(y.toString)
    }
  }

}

File: engine/runtime/src/main/java/org/enso/compiler/ImportExportCache.java
package org.enso.compiler;

import buildinfo.Info;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.oracle.truffle.api.TruffleFile;
import com.oracle.truffle.api.TruffleLogger;
import org.enso.compiler.data.BindingsMap;
import org.enso.editions.LibraryName;
import org.enso.interpreter.runtime.EnsoContext;
import org.enso.pkg.QualifiedName;
import org.enso.pkg.SourceFile;
import scala.collection.immutable.Map;
import scala.jdk.CollectionConverters;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.logging.Level;

public final class ImportExportCache extends Cache<ImportExportCache.CachedBindings, ImportExportCache.Metadata> {

    private final LibraryName libraryName;

    public ImportExportCache(LibraryName libraryName) {
        this.libraryName = libraryName;
        this.logLevel = Level.FINEST;
        this.stringRepr = libraryName.toString();
        this.entryName = libraryName.name();
        this.dataSuffix = bindingsCacheDataExtension;
        this.metadataSuffix = bindingsCacheMetadataExtension;
    }

    @Override
    protected byte[] metadata(String sourceDigest, String blobDigest, CachedBindings entry) {
        var mapper = new ObjectMapper();
        try {
            return mapper.writeValueAsString(new Metadata(sourceDigest, blobDigest)).getBytes(metadataCharset);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected CachedBindings validateReadObject(Object obj, Metadata meta, TruffleLogger logger) throws CacheException {
        if (obj instanceof MapToBindings bindings) {
            return new CachedBindings(libraryName, bindings, Optional.empty());
        } else {
            throw new CacheException("Expected ImportExportCache.FileToBindings, got " + obj.getClass());
        }
    }

    @Override
    protected Optional<Metadata> metadataFromBytes(byte[] bytes) {
        var maybeJsonString = new String(bytes, Cache.metadataCharset);
        var mapper = new ObjectMapper();
        try {
            return Optional.of(mapper.readValue(maybeJsonString, ImportExportCache.Metadata.class));
        } catch (JsonProcessingException e) {
            return Optional.empty();
        }
    }

    @Override
    protected Optional<String> computeDigest(CachedBindings entry, TruffleLogger logger) {
        return entry.sources().map(sources -> computeDigestOfLibrarySources(sources, logger));
    }
    @Override
    @SuppressWarnings("unchecked")
    protected Optional<String> computeDigestFromSource(EnsoContext context, TruffleLogger logger) {
        return context
                .getPackageRepository()
                .getPackageForLibraryJava(libraryName)
                .map(pkg -> computeDigestOfLibrarySources(pkg.listSourcesJava(), logger));
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Optional<Cache.Roots> getCacheRoots(EnsoContext context) {
        return context.getPackageRepository().getPackageForLibraryJava(libraryName).map(pkg -> {
            var bindingsCacheRoot =
                    pkg.getBindingsCacheRootForPackage(Info.ensoVersion());
            var localCacheRoot = bindingsCacheRoot.resolve(libraryName.namespace());
            var distribution = context.getDistributionManager();
            var pathSegments = CollectionConverters.ListHasAsScala(Arrays.asList(
                    pkg.namespace(),
                    pkg.name(),
                    pkg.config().version(),
                    Info.ensoVersion(),
                    libraryName.namespace()
            )).asScala();
            var path = distribution.LocallyInstalledDirectories().irCacheDirectory()
                    .resolve(pathSegments.mkString("/"));
            var globalCacheRoot = context.getTruffleFile(path.toFile());
            return new Cache.Roots(localCacheRoot, globalCacheRoot);
        });
    }

    @Override
    protected Object extractObjectToSerialize(CachedBindings entry) {
        return entry.bindings();
    }

    static class MapToBindings implements Serializable {
        private final Map<QualifiedName, BindingsMap> _entries;

        public MapToBindings(Map<QualifiedName, BindingsMap> entries) {
            this._entries = entries;
        }

        Map<QualifiedName, BindingsMap> entries() {
            return _entries;
        }
    }

    // CachedBindings is not a record **on purpose**. There appears to be a Frgaal bug leading to invalid compilation error.
    static class CachedBindings {
        private final LibraryName _libraryName;
        private final MapToBindings _bindings;
        private final Optional<List<SourceFile<TruffleFile>>> _sources;
        public CachedBindings(LibraryName libraryName, MapToBindings bindings, Optional<List<SourceFile<TruffleFile>>> sources) {
            this._libraryName = libraryName;
            this._bindings = bindings;
            this._sources = sources;
        }

        LibraryName libraryName() {
            return _libraryName;
        }
        MapToBindings bindings() {
            return _bindings;
        }
        Optional<List<SourceFile<TruffleFile>>> sources() {
            return _sources;
        }

    }

    public record Metadata(
            @JsonProperty("source_hash") String sourceHash,
            @JsonProperty("blob_hash") String blobHash) implements Cache.Metadata {}

    private static final String bindingsCacheDataExtension = ".bindings";
    private static final String bindingsCacheMetadataExtension =".bindings.meta";

}
